File: App/src/main/java/com/drownedinsound/ui/home/postList/BoardPostListPresenter.java
package com.drownedinsound.ui.home.postList;


import com.google.auto.factory.AutoFactory;
import com.google.auto.factory.Provided;

import com.drownedinsound.data.DisBoardRepo;
import com.drownedinsound.data.generatered.BoardPostList;
import com.drownedinsound.data.generatered.BoardPostSummary;
import com.drownedinsound.qualifiers.ForIoScheduler;
import com.drownedinsound.qualifiers.ForMainThreadScheduler;
import com.drownedinsound.ui.base.Navigator;

import java.util.Date;
import java.util.List;

import rx.Observer;
import rx.Scheduler;
import rx.Subscription;

@AutoFactory
public class BoardPostListPresenter implements BoardPostListContract.Presenter {

    private final BoardPostListContract.View boardListView;

    private final DisBoardRepo disBoardRepo;

    private final Scheduler mainThreadScheduler;

    private final Scheduler backgroundThreadScheduler;

    private final Navigator navigator;

    private Subscription loadingSubscription;

    public BoardPostListPresenter(
            BoardPostListContract.View boardListView,
            Navigator navigator,
            @Provided DisBoardRepo disBoardRepo,
            @Provided @ForMainThreadScheduler Scheduler mainThreadScheduler,
            @Provided @ForIoScheduler Scheduler backgroundThreadScheduler) {
        this.boardListView = boardListView;
        this.navigator = navigator;
        this.disBoardRepo = disBoardRepo;
        this.mainThreadScheduler = mainThreadScheduler;
        this.backgroundThreadScheduler = backgroundThreadScheduler;
    }


    @Override
    public void onViewCreated() {
        loadList(false);
        boardListView.setPresenter(this);
    }

    @Override
    public void onViewDisplayed() {

    }

    private void loadList(boolean forceUpdate) {
        if(loadingSubscription == null || loadingSubscription.isUnsubscribed()) {
            boardListView.showLoadingProgress(true);

            loadingSubscription = disBoardRepo
                    .getBoardPostSummaryList(boardListView.getBoardListType(), 0, forceUpdate)
                    .subscribeOn(backgroundThreadScheduler)
                    .observeOn(mainThreadScheduler)
                    .subscribe(new Observer<List<BoardPostSummary>>() {
                        @Override
                        public void onCompleted() {

                        }

                        @Override
                        public void onError(Throwable e) {
                            boardListView.showLoadingProgress(false);
                            boardListView.showErrorView();
                        }

                        @Override
                        public void onNext(List<BoardPostSummary> boardPostLists) {
                            boardListView.showBoardPostSummaries(boardPostLists);
                            boardListView.showLoadingProgress(false);
                        }
                    });
        }
    }


    @Override
    public void onViewHidden() {

    }

    private void stopLoading() {
        if (loadingSubscription != null) {
            loadingSubscription.unsubscribe();
            loadingSubscription = null;
        }
        boardListView.showLoadingProgress(false);
    }

    @Override
    public void onViewDestroyed() {
        stopLoading();
    }

    @Override
    public void handleRefresh() {
        loadList(true);
    }

    @Override
    public void handleBoardPostSelected(BoardPostSummary boardPostSummary) {
        @BoardPostList.BoardPostListType String boardListType
                = boardPostSummary.getBoardListTypeID();
        boardPostSummary.setLastViewedTime(new Date().getTime());
        boardPostSummary.setNumberOfTimesOpened(boardPostSummary.getNumberOfTimesOpened() + 1);

        disBoardRepo.setBoardPostSummary(boardPostSummary)
                .subscribeOn(backgroundThreadScheduler)
                .subscribe();

        navigator.showBoardPostScreen(boardListType,boardPostSummary.getBoardPostID());
    }


}

File: App/src/main/kotlin/com/drownedinsound/ui/home/postList/BoardPostListPresenter.kt
package com.drownedinsound.ui.home.postList

import com.drownedinsound.BoardPostSummaryModel
import com.drownedinsound.core.DisBoardsAppModule
import com.drownedinsound.data.DisBoardRepo
import com.drownedinsound.data.generatered.BoardPostList
import com.drownedinsound.data.generatered.BoardPostSummary
import com.drownedinsound.ui.base.Navigator
import com.google.auto.factory.AutoFactory
import com.google.auto.factory.Provided
import rx.Scheduler
import rx.Subscription
import timber.log.Timber
import java.util.*
import javax.inject.Named

@AutoFactory class BoardPostListPresenter(
        private val boardListView: BoardPostListContract.View,
        private val navigator: Navigator,
        @param:Provided private val boardPostMapper: BoardPostListModelMapper,
        @param:Provided private val disBoardRepo: DisBoardRepo,
        @param:Provided @param:Named(DisBoardsAppModule.MAIN_THREAD_SCHEDULER) private val mainThreadScheduler: Scheduler,
        @param:Provided @param:Named(DisBoardsAppModule.BACKGROUND_THREAD_SCHEDULER) private val backgroundThreadScheduler: Scheduler)
    : BoardPostListContract.Presenter {

    private var loadingSubscription: Subscription? = null
    private var boardPostSummaries: List<BoardPostSummary> = listOf()

    override fun onViewCreated() {
        loadList(false)
        boardListView.setPresenter(this)
    }

    override fun onViewDisplayed() = Unit

    private fun loadList(forceUpdate: Boolean) {
        if (loadingSubscription == null) {
            boardListView.showLoadingProgress(true)

            loadingSubscription = disBoardRepo
                    .getBoardPostSummaryList(boardListView.boardListType, 0, forceUpdate)
                    .map { boardPostSummaries -> storeAndMap(boardPostSummaries) }
                    .subscribeOn(backgroundThreadScheduler)
                    .observeOn(mainThreadScheduler)
                    .subscribe({ items -> handleItems(items) },
                            { error -> handleError(error) })
        }
    }

    private fun handleItems(boardPostSummaries: List<BoardPostSummaryModel>) {
        boardListView.showBoardPostSummaries(boardPostSummaries)
        boardListView.showLoadingProgress(false)
    }

    private fun handleError(e: Throwable) {
        Timber.d(e, "Error")
        boardListView.showLoadingProgress(false)
        boardListView.showErrorView()
    }

    private fun storeAndMap(boardPostSummaries: MutableList<BoardPostSummary>): List<BoardPostSummaryModel> {
        this.boardPostSummaries = boardPostSummaries
        return boardPostMapper.map(boardPostSummaries)
    }

    private fun stopLoading() {
        loadingSubscription?.unsubscribe()
        loadingSubscription = null
        boardListView.showLoadingProgress(false)
    }

    override fun onViewDestroyed() = stopLoading()

    override fun handleRefresh() = loadList(true)

    override fun handleBoardPostSelected(boardPostSummaryModel: BoardPostSummaryModel) {
        val summary = boardPostSummaries[boardPostSummaryModel.index]
        @BoardPostList.BoardPostListType val boardListType = summary.boardListTypeID
        summary.lastViewedTime = Date().time
        summary.numberOfTimesOpened = summary.numberOfTimesOpened + 1

        disBoardRepo.setBoardPostSummary(summary)
                .subscribeOn(backgroundThreadScheduler)
                .subscribe()

        navigator.showBoardPostScreen(boardListType, summary.boardPostID)
    }
}
